core/include/scipp/core/has_eval.h

```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+
+namespace scipp::core {
+
+template <class T> struct has_eval {
+  template <class U>
+  static auto test(int) -> decltype(std::declval<U>().eval(), std::true_type());
+  template <class> static std::false_type test(...);
+  static constexpr bool value =
+      std::is_same<decltype(test<T>(0)), std::true_type>::value;
```

Comment:
`value = decltype(test<T>(0))::value`?

Interesting solution by the way. Don't think I have seen this approach before.

Reply:
Hmm, I found a handful of solutions like this when googling how to detect this. Which approach are you used to?

I have used `boost` TTI before, but didn't want to bring in the heavy machinery for this.

Reply:
I am used to seeing solutions based on `std::enable_if` or `std::void_t` like 
```.cpp
template <class T, class=void>
struct supports_foo : std::false_type{};

template <class T>
struct supports_foo<T, void_t<decltype(std::declval<T>().get_foo())>>
: std::true_type{};
```
from https://blog.tartanllama.xyz/detection-idiom/

This has the (kind of academic) advantage that it does not mix template spcialisations and function overloading. Though the blog post does also mention the solution you used.
I have no objection to it, I was just a bit surprised.

Reply:
I switched to your proposal, since it seems slightly more readable.